.flowconfig

```diff
@@ -120,6 +129,7 @@ suppress_type=$FlowFixMeState
 
 module.file_ext=.js
 module.file_ext=.json
+module.file_ext=.android.js
 module.file_ext=.ios.js
```

Comment:
Hmm, from looking closer just now, I don't think I've totally nailed down the pattern. Some observations, though:

I've temporarily tried splitting `ZulipButton.js` into `ZulipButton.android.js` and `ZulipButton.ios.js`.

I notice that the invalid assertion `(123: string)` gets caught in both files—as long as this bit under `[ignore]` is removed first:

```diff
-; We fork some components by platform
-.*/*[.]android.js
```

With that intact, the error in the `.android.js` file isn't reported.

So that makes sense, I guess?

But then, I added an `android: boolean` prop to `ZulipButton.android.js`, and an `ios: boolean` prop to `ZulipButton.ios.js`, to see what the complaints were at `ZulipButton`'s callsites that obviously weren't passing either prop.

I got the error message from `ZulipButton.android.js`—`property android is missing in props`—when I tried with all these orderings:

```
module.file_ext=.json
module.file_ext=.js
module.file_ext=.android.js
module.file_ext=.ios.js
```

```
module.file_ext=.json
module.file_ext=.js
module.file_ext=.ios.js
module.file_ext=.android.js
```

```
module.file_ext=.json
module.file_ext=.android.js
module.file_ext=.ios.js
module.file_ext=.js
```

```
module.file_ext=.json
module.file_ext=.android.js
module.file_ext=.js
module.file_ext=.ios.js
```

```
module.file_ext=.json
module.file_ext=.ios.js
module.file_ext=.js
module.file_ext=.android.js
```

```
module.file_ext=.json
module.file_ext=.ios.js
module.file_ext=.android.js
module.file_ext=.js
```

Only when I commented out the `.android.js` line did I see the error message from `ZulipButton.ios.js`.

Reply:
So I think this is yet another thing pointing to generally avoiding `.android.js` and `.ios.js` files. My reason for adding the `module.file_ext=.android.js` line was because `react-native-paper` [reportedly](https://callstack.github.io/react-native-paper/getting-started.html#installation) requires it as part of its setup, and I thought this would be a good commit for the change to go in—but I didn't get to the point of testing that `react-native-paper` actually needs it, and we can do that at the time we install `react-native-paper` if we do.

For now I'll not add the `module.file_ext=.android.js` line. The `.*/*[.]android.js` cleanup seems OK though?

Reply:
> The `.*/*[.]android.js` cleanup seems OK though?

Hmm actually maybe better drop this, too; it's in the RN template app's Flow config. I'll probably thank myself when I'm doing a future RN upgrade.

Reply:
> But then, I added an `android: boolean` prop to `ZulipButton.android.js`, and an `ios: boolean` prop to `ZulipButton.ios.js`, to see what the complaints were at `ZulipButton`'s callsites that obviously weren't passing either prop.

Cool. From that data, I think a plausible hypothesis is:
* Flow will *check* each file that it finds at all. At this level, `foo.android.js` and `foo.js` are just two different files matching `*.js`; it doesn't care that one of the names has an extra `.` in it, and maybe doesn't even look at `module.file_ext`.
* But when checking some other file and *importing* a module `foo`, it consults `module.file_ext` to see what filenames could count: it wants to take the module name and append one of those extensions to get the filename.
* And if there are several extensions that lead to a valid file for the desired module name... well, apparently it doesn't care about the order the `module.file_ext` lines appear in the config. Maybe it just looks alphabetically? But then it'd be essential that both "android" and "ios" come before "js" in the alphabet... Or maybe it goes longest-first? That'd be a way of both ensuring `.whatever.js` comes before `.js`, and by happenstance putting `.android.js` before `.js`.

Docs don't seem to answer this question:
https://flow.org/en/docs/config/options/#toc-module-file-ext-string
but it's moot at least for now, so :shrug: .